core/src/main/java/org/ehcache/config/builders/ConfigurationBuilder.java

```diff
@@ -0,0 +1,208 @@
+      seed.getCacheConfigurations()), seed.getServiceCreationConfigurations()), seed.getClassLoader());
+  }
+
+  private ConfigurationBuilder() {
+    this.caches = emptyMap();
+    this.serviceConfigurations = emptyList();
+    this.classLoader = null;
+  }
+
+  private ConfigurationBuilder(ConfigurationBuilder builder, Map<String, CacheConfiguration<?, ?>> caches) {
+    this.caches = unmodifiableMap(caches);
+    this.serviceConfigurations = builder.serviceConfigurations;
+    this.classLoader = builder.classLoader;
+  }
+
+  private ConfigurationBuilder(ConfigurationBuilder builder, Collection<ServiceCreationConfiguration<?, ?>> serviceConfigurations) {
+    this.caches = builder.caches;
+    this.serviceConfigurations = unmodifiableCollection(serviceConfigurations);
+    this.classLoader = builder.classLoader;
+  }
+
+  private ConfigurationBuilder(ConfigurationBuilder builder, ClassLoader classLoader) {
+    this.caches = builder.caches;
+    this.serviceConfigurations = builder.serviceConfigurations;
+    this.classLoader = classLoader;
+  }
+
+  @Override
+  public Configuration build() {
+    return new DefaultConfiguration(caches, classLoader, serviceConfigurations.toArray(new ServiceCreationConfiguration<?, ?>[serviceConfigurations.size()]));
+  }
+
+  @Deprecated
+  public ConfigurationBuilder addCache(String alias, CacheConfiguration<?, ?> config) {
+    CacheConfiguration<?, ?> existing = getCache(alias);
+    if (existing == null) {
+      return withCache(alias, config);
+    } else {
+      throw new IllegalArgumentException("Cache '" + alias + "' already exists: " + existing);
+    }
+  }
+
+  @Deprecated
+  public ConfigurationBuilder removeCache(String alias) {
+    return withoutCache(alias);
+  }
+
+  @Override
+  public CacheConfiguration<?, ?> getCache(String alias) {
+    return caches.get(alias);
+  }
+
+  @Override
+  public ConfigurationBuilder withCache(String alias, CacheConfiguration<?, ?> config) {
+    Map<String, CacheConfiguration<?, ?>> newCaches = new HashMap<>(caches);
+    newCaches.put(alias, config);
+    return new ConfigurationBuilder(this, newCaches);
+  }
+
+  @Override
+  public ConfigurationBuilder withoutCache(String alias) {
+    Map<String, CacheConfiguration<?, ?>> newCaches = new HashMap<>(caches);
+    newCaches.remove(alias);
+    return new ConfigurationBuilder(this, newCaches);
+  }
+
+  @Override
+  public ConfigurationBuilder updateCache(String alias, UnaryOperator<FluentCacheConfigurationBuilder<?, ?, ?>> update) {
+    CacheConfiguration<?, ?> existing = getCache(alias);
+    if (existing == null) {
+      throw new IllegalArgumentException("Cache does not exist");
+    } else {
+      return withCache(alias, update.apply(existing.derive()).build());
+    }
+  }
+
+  @Deprecated
+  public ConfigurationBuilder addService(ServiceCreationConfiguration<?, ?> serviceConfiguration) {
```

Comment:
I presume that, even though this is a "new" class, it existed elsewhere before.  If there is existing non-internal code that uses this, it would benefit from having the `@deprecated` javadoc tag that indicates this method's replacement.

Reply:
Added javadoc.